LocalDispatcher improvements

johnhaddon opened the pull request:
This introduces a new LocalJobs dockable editor to replace the old floating local dispatcher window, initially just by moving the code from one place to another in https://github.com/GafferHQ/gaffer/commit/9223e2900e089c4506d0f6be698ce08eb74339ad. There then follows a long series of commits that improve the implementation of the LocalDispatcher to support improvements to the LocalJobs panel, with the main user-visible improvements being :

- The UI updates promptly using new signals provided by the Job class.
- Jobs stay in the listing until removed explicitly, rather than disappearing on completion.
- Task output is captured and shown in the Log widget.
- Progress messages are generally improved.
- More job properties are shown (start time, environment command etc).

murraystevenson commented:
Thanks John! I really like where this is headed.

A couple of thoughts from a user perspective:

It'd be nice to have some of the info in the Properties tab available as columns. Job start time & duration would be quite useful to have as columns when optimising renders. If possible, a peak memory column could be super useful too.

With the change to terminating background tasks at shutdown in 3dc1692, it's probably worth warning users on quit if they have active background tasks, much like we warn on unsaved script changes. I don't think we'd need to go so far as to allow them to quit while keeping the background tasks alive, but a reminder wouldn't hurt.

Currently all the task output captured from stdout in `handleOutput()` is reported as `INFO`. It'd be great if we could at least have warning messages from renderers reported as such so they gain the nice colour coding and filtering in the MessageWidget. Some cheap and cheerful pattern matching of the log output might cover the majority of cases?

johnhaddon commented:
I've pushed a rebase to pick up the `psutil` module from #5577. All the original commits (prior to and including https://github.com/GafferHQ/gaffer/pull/5557/commits/ca0dc8fa7011638826dd7f5a3b4e67153fc12089) are as they were before except for a couple of minor changes that it made more sense to squash in :

- The MessageHandler in https://github.com/GafferHQ/gaffer/pull/5557/commits/5a24438dc9b0196ba07fe83c50971fa2444f051c now splits messages on `\n`, for improved formatting in the log widget.
- Removed the splitting by line for formatted exceptions in https://github.com/GafferHQ/gaffer/pull/5557/commits/3805dd4362a331fefa0ff6e0437853d9d91b1de6, since that is now done by our MessageHandler and by the OStreamMessageHandler in Cortex following https://github.com/ImageEngine/cortex/pull/1399.

> It'd be nice to have some of the info in the Properties tab available as columns. Job start time & duration would be quite useful to have as columns when optimising renders. If possible, a peak memory column could be super useful too.
 
I added start time and duration in https://github.com/GafferHQ/gaffer/pull/5557/commits/74505372bec917fd4587a1931ac72bd8f2283983 and https://github.com/GafferHQ/gaffer/pull/5557/commits/3b084cd7671f1ef27581255b686ce661290c658e. Peak memory isn't available in a cross-platform way from `psutil`, and the method we use for querying on Linux in the `stats` app only seems to apply to the current process (or all its children). We could continually poll for that, but I'd like to limit the overhead of LocalDispatcher as long as the LocalJobs window is hidden. So I've held off on that for now.

On the topic of overhead, I also discovered that the overhead of `Job.statistics()` was completely unreasonable if you have more than a few jobs running at once. I fixed that in https://github.com/GafferHQ/gaffer/pull/5557/commits/d3773b95d0688ada5e842c96b02b2aefcc2daa05, which shold also fix the stats on Windows.

I also added the improved PathColumn sorting options we discussed in https://github.com/GafferHQ/gaffer/pull/5557/commits/f9531b43cfb11299a9c9113fcaa889b7ce61fe89 and used it to add sorting by status in https://github.com/GafferHQ/gaffer/pull/5557/commits/577daa236fd7a7e70434916570aa5f71365d1ca2.

> Currently all the task output captured from stdout in handleOutput() is reported as INFO. It'd be great if we could at least have warning messages from renderers reported as such so they gain the nice colour coding and filtering in the MessageWidget. Some cheap and cheerful pattern matching of the log output might cover the majority of cases?

Done in https://github.com/GafferHQ/gaffer/pull/5557/commits/39d383f4d69184a047afad1b0b4e4a624d126cf3.

> With the change to terminating background tasks at shutdown, it's probably worth warning users on quit if they have active background tasks, much like we warn on unsaved script changes.

Good point. I haven't done this yet because it raises a few other questions about how we handle shutdown - I'll bring those up when we catch up later today. Nevertheless, I would like to get the rest merged now if the new changes look good to you. This is a big PR already, and I'd benefit from the headspace a merge and a clean slate would give me.

murraystevenson commented on python/GafferDispatchUI/LocalJobs.py:
```diff
@@ -0,0 +1,365 @@
+
+		return GafferUI.PathColumn.CellData(
+			value = "{:.2f}GB".format( memory / (1024 ** 3) ) if memory is not None else "---",
+			sortValue = memory if memory is not None else 0,
```
I've seen this lead to `OverflowError: bad numeric conversion: positive overflow` on batches with memory usage above 2 gigabytes, maybe we should sort by the remapped to gigabytes value instead?

johnhaddon replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/5557/commits/97b829a7ef9fff7ae75181ccc0992d10c97cc4ad by returning UInt64Data explicitly rather than relying on the implicit conversion to IntData that was overflowing.

murraystevenson commented on python/GafferDispatchUI/LocalJobs.py:
```diff
@@ -0,0 +1,365 @@
+						_StatusColumn(),
+						GafferUI.PathListingWidget.StandardColumn( "Name", "localDispatcher:jobName", sizeMode = GafferUI.PathColumn.SizeMode.Stretch ),
+						GafferUI.PathListingWidget.StandardColumn( "Id", "localDispatcher:id" ),
+						GafferUI.PathListingWidget.StandardColumn( "Start Time", "localDispatcher:startTime" ),
```
The start time column is displaying oddly for me. It appears to be considering my local time as if it were UTC and then displays it with my local timezone offset from UTC applied, so the displayed start time is 8 hours earlier than my system clock. For example, if a job started at 9pm PST, the start time column will show 1pm. The start time listed in the Properties tab is correct, it's just the column display that is wrong.

I've tried this on Windows as well, with the same outcome.

If I change my system timezone to GMT/UTC then the start time column correctly matches my system clock which would explain why you wouldn't have run into this, though you should be able to reproduce it by shifting your own system time zone away from GMT...

johnhaddon replied:
Following Great Britain's glorious exit from the European Union, Gaffer will be representing time solely as god intended, in GMT. I'm sure you will appreciate the simplification and reduction in red tape this brings.

Fixed in https://github.com/GafferHQ/gaffer/pull/5557/commits/62ab2167c8fb055fa40a9cf221feb1b49b2994a2 :)

murraystevenson reviewed (commented):
Thanks John, these improvements are great! I've made two comments inline, one should be super straightforward to fix and hopefully the other isn't a can of time-travelling worms. I think if we sort those we'll be good to merge and move on...

johnhaddon commented:
Pushed another rebase with additional fixes - only the last two commits are new, everything else is as before.

johnhaddon commented:
Pushed one more fix - https://github.com/GafferHQ/gaffer/pull/5557/commits/a1f01c79716c52088398cf30760c4043c04626a1.

murraystevenson reviewed (approved):
Thanks for those last fixes, and for taking the correct stance on GMT and its place worldwide - that'll make scheduling meetings far-far more straightforward. Now if there's only something we can do about daylight saving... :sunrise: 

Looks good to me to squash, resolve the Changes.md conflict and merge!
